packages/SwingSet/src/kernel/virtualObjectManager.js

```diff
@@ -0,0 +1,342 @@
+          );
+          syscall.vatstoreSet(vkey, JSON.stringify(m.serialize(harden(value))));
+        } else {
+          assertKeyExists(vkey);
+          backingMap.set(key, value);
+        }
+      },
+    });
+  }
+
+  /**
+   * Fetch an object's state from secondary storage.
+   *
+   * @param {string} instanceKey  The instance ID of the object whose state is
+   *    being fetched.
+   *
+   * @returns {*} an object representing the object's stored state.
+   */
+  function fetch(instanceKey) {
+    return JSON.parse(syscall.vatstoreGet(instanceKey));
+  }
+
+  /**
+   * Write an object's state to secondary storage.
+   *
+   * @param {string} instanceKey  The instance ID of the object whose state is
+   *    being stored.
+   * @param {*} rawData  A data object representing the state to be written.
+   */
+  function store(instanceKey, rawData) {
+    syscall.vatstoreSet(instanceKey, JSON.stringify(rawData));
+  }
+
+  /**
+   * Make a new kind of virtual object.
+   *
+   * @param {*} instanceMaker  A function of the form `instanceMaker(state)` that
+   *    will return a representative instance wrapped around the given state.
+   *
+   * @returns {*} a maker function that can be called to manufacture new
+   *    instance of this kind of object.  The parameters of the maker function
+   *    are those of the `initialize` method implemented in the representative
+   *    produced by the `instanceMaker` parameter.
+   */
+  function makeKind(instanceMaker) {
+    const kindID = `${allocateExportID()}`;
+    let nextInstanceID = 1;
+
+    function makeInnerSelf(instanceKey, rawData) {
+      const innerSelf = {
+        prev: undefined,
+        next: undefined,
+        instanceKey,
+        rawData,
+        dropData() {
+          store(instanceKey, innerSelf.rawData);
+          innerSelf.rawData = null;
+        },
+      };
+      return innerSelf;
+    }
+
+    function makeRepresentative(innerSelf, initArgs) {
+      function ensureState() {
+        if (!innerSelf.rawData) {
+          const instanceKey = innerSelf.instanceKey;
+          innerSelf = cache.lookup(instanceKey);
+          if (!innerSelf) {
+            innerSelf = makeInnerSelf(instanceKey, fetch(instanceKey));
+            cache.remember(innerSelf);
+          }
+        }
+        cache.refresh(innerSelf);
+      }
+
+      function wrapData() {
+        innerSelf.activeData = {};
+        for (const prop of Object.getOwnPropertyNames(innerSelf.rawData)) {
+          Object.defineProperty(innerSelf.activeData, prop, {
+            get: () => {
+              ensureState();
+              return m.unserialize(innerSelf.rawData[prop]);
+            },
+            set: value => {
+              ensureState();
+              innerSelf.rawData[prop] = m.serialize(value);
+            },
+          });
+        }
+        return harden(innerSelf.activeData);
+      }
+
+      if (!innerSelf.rawData && initArgs) {
```

Comment:
If this is only being used for initialization, I think it should move outside `makeRepresentative` and into `makeNewInstance`

Reply:
Hmm.  Lemme noodle on that one.  It would make the separation of concerns between initialization and reload a bit clearer.

Reply:
Done.